File: tests/BaseTestCase.php
<?php

namespace RocketGate\Sdk\Tests;

use PHPUnit\Framework\TestCase;
use RocketGate\Sdk\GatewayRequest;
use RocketGate\Sdk\GatewayResponse;
use RocketGate\Sdk\GatewayService;

abstract class BaseTestCase extends TestCase
{
    /**
     * @var GatewayService
     */
    protected $service;

    /**
     * @var GatewayRequest
     */
    protected $request;

    /**
     * @var GatewayResponse
     */
    protected $response;

    /**
     * @var int
     */
    protected $merchantId = 1;

    /**
     * @var string
     */
    protected $merchantPassword = 'testpassword';

    /**
     * @var string
     */
    protected $customerId;

    /**
     * @var string
     */
    protected $invoiceId;

    /**
     * {@inheritdoc}
     */
    protected function setUp()
    {
        parent::setUp();

        $this->service  = new GatewayService(true);
        $this->response = new GatewayResponse();
        $this->request  = new GatewayRequest();

        // Merchant data
        $this->request->Set(GatewayRequest::MERCHANT_ID(), $this->merchantId);
        $this->request->Set(GatewayRequest::MERCHANT_PASSWORD(), $this->merchantPassword);

        // Customer data
        $this->request->Set(GatewayRequest::CUSTOMER_FIRSTNAME(), "Joe");
        $this->request->Set(GatewayRequest::CUSTOMER_LASTNAME(), "PHPTester");
        $this->request->Set(GatewayRequest::EMAIL(), "phptest@fakedomain.com");

        // Credit card data
        $this->request->Set(GatewayRequest::CARDNO(), "4111111111111111");
        $this->request->Set(GatewayRequest::EXPIRE_MONTH(), "02");
        $this->request->Set(GatewayRequest::EXPIRE_YEAR(), "2010");
        $this->request->Set(GatewayRequest::CVV2(), "999");

        $time = time();
        $this->customerId = $time . '.PHPTest';
        $this->invoiceId  = $time . '.' . $this->getTestName();

        $this->request->Set(GatewayRequest::MERCHANT_CUSTOMER_ID(), $this->customerId);
        $this->request->Set(GatewayRequest::MERCHANT_INVOICE_ID(), $this->invoiceId);
    }

    abstract function getTestName() : string;

    function getService() : GatewayService
    {
        return clone $this->service;
    }

    function getRequest() : GatewayRequest
    {
        return clone $this->request;
    }

    function getResponse() : GatewayResponse
    {
        return clone $this->response;
    }
}

File: tests/BaseTestCase.rb
require 'minitest/autorun'
require_relative '../GatewayService'
require_relative '../GatewayRequest'
require_relative '../GatewayResponse'

module RocketGate
  class BaseTestCase < Minitest::Test

    def setup
      @merchantId = 1

      @merchantPassword = 'testpassword'

      @service = GatewayService.new
      @service.SetTestMode true

      @response = GatewayResponse.new
      @request = GatewayRequest.new

      # Merchant data
      @request.Set(GatewayRequest::MERCHANT_ID, @merchantId)
      @request.Set(GatewayRequest::MERCHANT_PASSWORD, @merchantPassword)

      # Customer data
      @request.Set(GatewayRequest::CUSTOMER_FIRSTNAME, "Joe")
      @request.Set(GatewayRequest::CUSTOMER_LASTNAME, "RubyTester")
      @request.Set(GatewayRequest::EMAIL, "rubytest@fakedomain.com")

      # Credit card data
      @request.Set(GatewayRequest::CARDNO, "4111111111111111")
      @request.Set(GatewayRequest::EXPIRE_MONTH, "02")
      @request.Set(GatewayRequest::EXPIRE_YEAR, "2010")
      @request.Set(GatewayRequest::CVV2, "999")

      t = Time.now.to_i.to_s

      @customerId = t + '.RubyTest'
      @invoiceId = t + '.' + get_test_name

      @request.Set(GatewayRequest::MERCHANT_CUSTOMER_ID, @customerId)
      @request.Set(GatewayRequest::MERCHANT_INVOICE_ID, @invoiceId)
    end

    def print_resp(response)
      puts "Response Code: " + response.Get(GatewayResponse::RESPONSE_CODE)
      puts "Reason Code: " + response.Get(GatewayResponse::REASON_CODE)
      puts "Exception: " + response.Get(GatewayResponse::EXCEPTION)
      puts "GUID: " + response.Get(GatewayResponse::TRANSACT_ID)
      puts "AuthNo: " + response.Get(GatewayResponse::AUTH_NO)
      puts "AVS: " + response.Get(GatewayResponse::AVS_RESPONSE)
      puts "CVV2: " + response.Get(GatewayResponse::CVV2_CODE)
      puts "CardHash: " + response.Get(GatewayResponse::CARD_HASH)
      puts "Account: " + response.Get(GatewayResponse::MERCHANT_ACCOUNT)
      puts "Scrub: " + response.Get(GatewayResponse::SCRUB_RESULTS)
    end

    def get_test_name
      raise 'implement'
    end
  end
end
